test/e2e/scheduling/predicates.go

```diff
@@ -724,6 +727,231 @@ var _ = SIGDescribe("SchedulerPredicates [Serial]", func() {
+			ginkgo.By("make the nodes have balanced resource usage ratio")
+			err := createBalancedPodForNodes(ctx, f, cs, ns, nodeList.Items, podRequestedResource, 0.5)
+			framework.ExpectNoError(err)
+
+			k := v1.LabelHostname
+			// All of Pods hate each other and get scheduled in a different node.
+			affinity := &v1.Affinity{
+				PodAntiAffinity: &v1.PodAntiAffinity{
+					RequiredDuringSchedulingIgnoredDuringExecution: []v1.PodAffinityTerm{
+						{
+							LabelSelector: &metav1.LabelSelector{
+								MatchExpressions: []metav1.LabelSelectorRequirement{
+									{
+										Key:      "security",
+										Operator: metav1.LabelSelectorOpIn,
+										Values:   []string{"S1"},
+									},
+								},
+							},
+							TopologyKey:    k,
+							Namespaces:     []string{ns},
+							MatchLabelKeys: []string{"pod-template-hash"},
+						},
+					},
+				},
+			}
+
+			intstrPtr := func(i int) *intstr.IntOrString {
+				num := intstr.FromInt(i)
+				return &num
+			}
+			deploymentPodLabels := map[string]string{"test": "matchLabelKey1", "security": "S1"}
+			deploymentName := "test-matchlabelkey-deployment"
+			deploymentReplicas := int32(numNode)
+			deploymentImage := imageutils.GetPauseImageName()
+			deploymentStrategyType := appsv1.RollingUpdateDeploymentStrategyType
+			newDeployment := e2edeployment.NewDeployment(deploymentName, deploymentReplicas, deploymentPodLabels, "pause", deploymentImage, deploymentStrategyType)
+			newDeployment.Spec.Strategy.RollingUpdate = &appsv1.RollingUpdateDeployment{
+				MaxUnavailable: intstrPtr(0),
+				MaxSurge:       intstrPtr(numNode),
+			}
+			newDeployment.Spec.Template.Spec.Affinity = affinity
+
+			d, err := cs.AppsV1().Deployments(ns).Create(ctx, newDeployment, metav1.CreateOptions{})
+			framework.ExpectNoError(err)
+
+			framework.Logf("Make sure deployment %q is complete", deploymentName)
+			pollInterval := 1 * time.Second
+			err = testutils.WaitForDeploymentCompleteAndCheckRolling(cs, d, framework.Logf, pollInterval, wait.ForeverTestTimeout)
+			framework.ExpectNoError(err)
+
+			framework.Logf("Make sure all Pods in deployment %q is scheduled", deploymentName)
+			err = wait.Poll(pollInterval, wait.ForeverTestTimeout, deploymentReplicasScheduled(ctx, cs, ns, deploymentPodLabels))
+			framework.ExpectNoError(err)
+
+			framework.Logf("Make sure all Pods in deployment %q is scheduled on different Node", deploymentName)
+			nodenames, err := deploymentReplicasNodeNames(ctx, cs, ns, deploymentPodLabels)
+			framework.ExpectNoError(err)
+			nodenameset := sets.Set[string]{}
+			for _, nodename := range nodenames {
+				nodenameset.Insert(nodename)
+			}
+			framework.ExpectEqual(len(nodenames), nodenameset.Len())
+
+			framework.Logf("update deployment %q to run rolling upgrade", deploymentName)
+			// change label to kick the rolling upgrade.
+			d.Spec.Template.Labels["test"] = "matchLabelKey2"
+			d, err = cs.AppsV1().Deployments(ns).Update(ctx, newDeployment, metav1.UpdateOptions{})
```

Comment:
nit: Update() in real-world case, may meet API confliction very easily (as long as ResourceVersion doesn't match). Maybe change to retriable Update() or Patch().